File: src/mvmctl/cli/kernel.py
"""Kernel management commands."""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING, Any, cast

import typer
from rich.console import Console

from mvmctl.api import KernelImportInput as _KernelImportInput
from mvmctl.api import KernelInput as _KernelInput
from mvmctl.api import KernelOperation as _KernelOperation
from mvmctl.api import KernelPullInput as _KernelPullInput

if TYPE_CHECKING:
    from mvmctl.api.inputs._kernel_import_input import KernelImportInput
    from mvmctl.api.inputs._kernel_input import KernelInput
    from mvmctl.api.inputs._kernel_pull_input import KernelPullInput
    from mvmctl.api.kernel_operations import KernelOperation
else:
    KernelOperation = _KernelOperation
    KernelPullInput = _KernelPullInput
    KernelInput = _KernelInput
    KernelImportInput = _KernelImportInput
from mvmctl.cli._completion import _complete_kernel_ids
from mvmctl.models import KernelItem, VersionInfo
from mvmctl.models.result import OperationResult, ProgressEvent
from mvmctl.utils.cli import handle_errors, mvm_cli

kernel_app = typer.Typer(
    help="Kernel management",
    no_args_is_help=True,
    add_completion=False,
)


@kernel_app.callback()
def kernel_callback(ctx: typer.Context) -> None:
    pass


@kernel_app.command(name="ls")
@handle_errors
def kernel_ls(
    json_output: bool = typer.Option(False, "--json", help="Output as JSON"),
    remote: bool = typer.Option(
        False, "--remote", "-r", help="Show available remote kernel versions"
    ),
    no_cache: bool = typer.Option(
        False,
        "--no-cache",
        help="Skip cached version listing and fetch live from upstream",
    ),
) -> None:
    """List cached kernels (or available remote kernels with --remote)."""
    if remote:
        with Console().status("Fetching remote kernel versions"):
            versions = cast(
                list[VersionInfo],
                KernelOperation.list_all(remote=True, no_cache=no_cache),
            )
        _list_remote_kernels(versions, json_output=json_output)
    else:
        kernels = cast(
            list[KernelItem],
            KernelOperation.list_all(),
        )

        if json_output:
            data = [KernelOperation._kernel_to_dict(k) for k in kernels]
            typer.echo(json.dumps(data, indent=2, default=str))
            return

        rows: list[list[str]] = []
        for k in kernels:
            rows.append(
                [
                    mvm_cli.format_marker(k.is_default),
                    mvm_cli.format_id(k.id),
                    mvm_cli.format_name(k.base_name, not k.is_present),
                    k.version,
                    k.arch,
                    k.type,
                    mvm_cli.format_timestamp(k.created_at),
                ]
            )

        mvm_cli.table(
            columns=["", "ID", "Name", "Version", "Arch", "Type", "Added"],
            rows=rows,
        )


def _list_remote_kernels(
    versions: list[VersionInfo], *, json_output: bool
) -> None:
    """Render remote available kernels grouped by type."""
    if json_output:
        data: list[dict[str, Any]] = [
            {
                "version": v.version,
                "type": v.type,
                "display_name": v.display_name,
                "download_url": v.download_url,
                "sha256_url": v.sha256_url,
                "format": v.format,
            }
            for v in versions
        ]
        typer.echo(json.dumps(data, indent=2))
        return

    # Group by type
    groups: dict[str, list[VersionInfo]] = {}
    for v in versions:
        groups.setdefault(v.type, []).append(v)

    sorted_types = sorted(groups.keys())

    rows: list[list[str]] = []
    for type_key in sorted_types:
        version_list = groups[type_key]
        if not version_list:
            continue

        parts = type_key.split("-", maxsplit=1)
        type_display = (
            f"{parts[0].title()} {parts[1]}"
            if len(parts) > 1
            else type_key.title()
        )
        suffix = " (build required)" if type_key.startswith("official") else ""
        rows.append([type_key, f"{type_display}{suffix}"])

        for j, v in enumerate(version_list):
            is_last = j == len(version_list) - 1
            prefix = "  └─ " if is_last else "  ├─ "
            display = v.display_name or v.version
            rows.append([f"{prefix}{v.version}", display])

    if not rows:
        mvm_cli.info("No remote kernels available.")
        return

    mvm_cli.table(
        columns=["Type / Version", "Description"],
        rows=rows,
    )


@kernel_app.command(name="inspect")
@handle_errors
def kernel_inspect(
    prefix: str = typer.Argument(
        ...,
        help="Kernel ID prefix to inspect",
        autocompletion=_complete_kernel_ids,
    ),
    json_output: bool = typer.Option(False, "--json", help="Output as JSON"),
) -> None:
    """Show detailed information about a kernel."""
    info = KernelOperation.inspect(KernelInput(id=[prefix]))

    if json_output:
        typer.echo(json.dumps(info, indent=2, default=str))
        return

    name = info.get("kernel", {}).get("name", prefix)
    mvm_cli.print_dict_tree(info, title=f"Kernel: {name}")


@kernel_app.command(name="pull")
@handle_errors
def kernel_pull(
    kernel_selector: str | None = typer.Argument(
        None,
        help="Shorthand: type:version (e.g. official:6.19.9). "
        "Use '--type' and '--version' options for explicit control.",
        autocompletion=_complete_kernel_ids,
    ),
    kernel_type: str | None = typer.Option(
        None, "--type", help="Kernel type: firecracker or official"
    ),
    version: str | None = typer.Option(
        None, "--version", help="Kernel version"
    ),
    arch: str | None = typer.Option(
        None, "--arch", help="Architecture (x86_64, arm64)"
    ),
    set_default: bool = typer.Option(
        False, "--default", "-d", help="Set as default after fetch"
    ),
    jobs: int | None = typer.Option(
        None, "--jobs", help="Parallel build jobs (official only)"
    ),
    keep_build_dir: bool = typer.Option(
        False, "--keep-build-dir", help="Keep build directory (official only)"
    ),
    clean_build: bool = typer.Option(
        False, "--clean-build", help="Skip cache (official only)"
    ),
    kernel_config: Path | None = typer.Option(
        None,
        "--config",
        help="Custom kernel config file to apply as a fragment",
    ),
    features: str | None = typer.Option(
        None,
        "--features",
        help="Comma-separated kernel features (kvm, nftables, tuntap)",
    ),
) -> None:
    """Pull or build a kernel.

    Examples:
        mvm kernel pull official:6.19.9
        mvm kernel pull official:6.19.9 --default
        mvm kernel pull --type official --version 6.19.9
        mvm kernel pull firecracker --arch arm64
    """
    # Parse ``type:version`` shorthand syntax (e.g. ``official:6.19.9``)
    effective_type = kernel_type
    effective_version = version
    if kernel_type is None and kernel_selector is not None:
        if ":" in kernel_selector:
            parts = kernel_selector.rsplit(":", maxsplit=1)
            effective_type = parts[0]
            effective_version = parts[1]
        else:
            effective_type = kernel_selector

    if effective_type is None:
        mvm_cli.error(
            "Kernel type is required. "
            "Use 'mvm kernel pull --type official' or "
            "'mvm kernel pull official:6.19.9'"
        )
        raise typer.Exit(code=1)

    inputs = KernelPullInput(
        kernel_type=effective_type,
        version=effective_version,
        arch=arch,
        jobs=jobs,
        keep_build_dir=keep_build_dir,
        clean_build=clean_build,
        kernel_config=kernel_config,
        set_default=set_default,
        features=features or "",
    )
    console = Console()
    with console.status("", spinner="dots") as status:

        def _on_progress(event: ProgressEvent) -> None:
            if event.message:
                status.update(event.message)

        result = KernelOperation.pull(inputs, on_progress=_on_progress)
    if isinstance(result, OperationResult):
        if result.is_error:
            mvm_cli.error(result.message)
            raise typer.Exit(code=1)
        if result.status == "skipped":
            mvm_cli.info(result.message)
            if result.item:
                mvm_cli.info(f"  ID: {mvm_cli.format_id(result.item.id)}")
            raise typer.Exit(code=0)
        if result.item:
            mvm_cli.success(
                f"Pulled: {result.item.name} "
                f"(ID: {mvm_cli.format_id(result.item.id)})"
            )
            resolved_features = (result.metadata or {}).get("features", [])
            if resolved_features:
                mvm_cli.info(
                    "Enabled features: " + ", ".join(resolved_features)
                )
    else:
        # Fallback for unexpected non-OperationResult returns
        mvm_cli.success("Pull completed")


@kernel_app.command(
    name="default",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
@handle_errors
def kernel_set_default(
    ctx: typer.Context,
    kernel_id: str = typer.Argument(
        None,
        help="Kernel ID prefix or name",
        autocompletion=_complete_kernel_ids,
    ),
) -> None:
    """Set a kernel as the default."""
    kernel_id = mvm_cli.check_name_arg(ctx, kernel_id)
    inputs = KernelInput(id=[kernel_id])
    result = KernelOperation.set_default(inputs)

    if result.is_error:
        mvm_cli.error(result.message)
        raise typer.Exit(code=1)

    mvm_cli.success(result.message or f"Default kernel set to: {kernel_id}")


@kernel_app.command(
    name="rm",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
@handle_errors
def kernel_rm(
    ctx: typer.Context,
    identifiers: list[str] = typer.Argument(
        None,
        help="Kernel ID prefixes or names to remove",
        autocompletion=_complete_kernel_ids,
    ),
    force: bool = typer.Option(
        False, "--force", "-f", help="Remove even if referenced by VMs"
    ),
) -> None:
    """Remove one or more kernels."""
    effective_ids: list[str] = list(identifiers) if identifiers else []
    if not effective_ids:
        mvm_cli.error("Provide at least one kernel ID or name")
        raise typer.Exit(code=1)

    inputs = KernelInput(id=effective_ids, force=force)
    batch_result = KernelOperation.remove(inputs)

    for item_result in batch_result.items:
        if item_result.is_ok:
            mvm_cli.success(item_result.message or "Removed")
        else:
            mvm_cli.error(item_result.message or "Remove failed")

    if batch_result.has_any_error:
        raise typer.Exit(code=1)


@kernel_app.command(name="import")
@handle_errors
def kernel_import(
    name: str = typer.Argument(
        ...,
        help="User-assigned name for this kernel entry",
    ),
    path: Path = typer.Argument(
        ...,
        help="Path to vmlinux file",
        exists=True,
        readable=True,
    ),
    version: str | None = typer.Option(
        None,
        "--version",
        help="Override auto-detected kernel version",
    ),
    arch: str | None = typer.Option(
        None,
        "--arch",
        help="Kernel architecture (default: auto-detected from filename or platform)",
    ),
    set_default: bool = typer.Option(
        False, "--default", "-d", help="Set as default after import"
    ),
) -> None:
    """Register a vmlinux file as a kernel in the database.

    Examples:

        mvm kernel import my-kernel ./vmlinux-6.1-x86_64

        mvm kernel import my-kernel ./vmlinux-custom --version 6.1 --arch x86_64 --default
    """
    if not path.exists():
        mvm_cli.error(f"Source file not found: {path}")
        raise typer.Exit(code=1)

    inputs = KernelImportInput(
        name=name,
        path=path,
        version=version,
        arch=arch,
        set_default=set_default,
    )
    result = KernelOperation.import_(inputs)

    if result.is_error:
        mvm_cli.error(result.message or f"Import failed: {name}")
        raise typer.Exit(code=1)

    assert result.item is not None
    short_id = mvm_cli.format_id(result.item.id)
    mvm_cli.success(f"Imported: {result.item.name}")
    mvm_cli.info(f"  ID:   {short_id}")

    if set_default:
        mvm_cli.success(f"Default kernel set to: {name}")

    raise typer.Exit(code=0)


__all__ = ["kernel_app"]

File: internal/cli/kernel.go
// Package cli — kernel management commands
package cli

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"mvmctl/internal/cli/common"
	"mvmctl/internal/infra/event"
	"mvmctl/internal/lib/model"
	"mvmctl/pkg/api"
	"mvmctl/pkg/api/inputs"

	"github.com/spf13/cobra"
)

// kernelColumns defines the local listing columns for kernels.
var kernelColumns = []common.ListingColumn{
	{Header: "", Extract: func(v any) string { return common.Cli.FormatMarker(v.(*model.KernelItem).IsDefault) }},
	{Header: "ID", Extract: func(v any) string { return common.Cli.FormatID(v.(*model.KernelItem).ID) }},
	{Header: "Name", Extract: func(v any) string {
		return common.Cli.FormatName(v.(*model.KernelItem).BaseName, !v.(*model.KernelItem).IsPresent)
	}},
	{Header: "Version", Extract: func(v any) string { return v.(*model.KernelItem).Version }},
	{Header: "Type", Extract: func(v any) string { return v.(*model.KernelItem).Type }},
	{Header: "Arch", Extract: func(v any) string { return v.(*model.KernelItem).Arch }, LongOnly: true},
	{
		Header:  "Created",
		Extract: func(v any) string { return common.Cli.FormatTimestamp(v.(*model.KernelItem).CreatedAt, "relative") },
	},
}

func NewKernelCmd(kernelAPI api.KernelAPI, configAPI api.ConfigAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "kernel",
		Short: "Kernel management",
		Long:  "Manage kernels — list, pull, remove, inspect, set default, import.",
	}

	cmd.AddCommand(newKernelListCmd(kernelAPI, configAPI))
	cmd.AddCommand(newKernelPullCmd(kernelAPI))
	cmd.AddCommand(newKernelRemoveCmd(kernelAPI))
	cmd.AddCommand(newKernelInspectCmd(kernelAPI))
	cmd.AddCommand(newKernelDefaultCmd(kernelAPI))
	cmd.AddCommand(newKernelImportCmd(kernelAPI))

	return cmd
}

func newKernelListCmd(kernelAPI api.KernelAPI, configAPI api.ConfigAPI) *cobra.Command {
	var jsonOutput bool
	var longOutput bool
	var remote bool
	var noCache bool

	cmd := &cobra.Command{
		Use:     "ls",
		Aliases: []string{"list"},
		Short:   "List cached kernels (or available remote kernels with --remote)",
		RunE: func(cmd *cobra.Command, args []string) error {
			if remote {
				fmt.Fprintln(os.Stderr, "Fetching remote kernel versions...")
				_, remoteVersions, err := kernelAPI.KernelList(cmd.Context(), true, noCache, nil)
				if err != nil {
					return err
				}

				if jsonOutput {
					if remoteVersions == nil {
						remoteVersions = []model.VersionInfo{}
					}
					b, _ := json.MarshalIndent(remoteVersions, "", "  ")
					fmt.Println(string(b))
					return nil
				}

				if len(remoteVersions) == 0 {
					fmt.Println("No remote kernels available.")
					return nil
				}

				common.RenderVersionTree(remoteVersions)
				return nil
			}

			// Local listing
			kernels, _, err := kernelAPI.KernelList(cmd.Context(), false, false, nil)
			if err != nil {
				return err
			}

			if jsonOutput {
				if kernels == nil {
					kernels = []*model.KernelItem{}
				}
				b, _ := json.MarshalIndent(kernels, "", "  ")
				fmt.Println(string(b))
				return nil
			}

			// Local listing
			style := common.Cli.ResolveListingStyle(cmd.Context(), configAPI, longOutput)
			common.RenderListing(kernels, kernelColumns, style)
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")
	cmd.Flags().BoolVar(&longOutput, "long", false, "Show full listing with all columns")
	cmd.Flags().BoolVarP(&remote, "remote", "r", false, "Show available remote kernel versions")
	cmd.Flags().BoolVar(&noCache, "no-cache", false, "Skip cached version listing and fetch live from upstream")

	return cmd
}

func newKernelPullCmd(kernelAPI api.KernelAPI) *cobra.Command {
	var kernelType string
	var version string
	var setDefault bool
	var jobs int
	var keepBuildDir bool
	var cleanBuild bool
	var kernelConfig string
	var features string

	cmd := &cobra.Command{
		Use:               "pull [type:version]",
		Short:             "Pull or build a kernel",
		ValidArgsFunction: completeKernelIDs,
		Long: `Pull or build a kernel.

Examples:
  mvm kernel pull official:6.19.9
  mvm kernel pull official:6.19.9 --default
  mvm kernel pull --type official --version 6.19.9`,
		Args: cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			effectiveType := kernelType
			effectiveVersion := version

			if len(args) > 0 {
				selector := args[0]
				if strings.Contains(selector, ":") {
					// Split from the right (last colon) so "a:b:c" → type="a:b", version="c"
					// so "a:b:c" → type="a:b", version="c"
					idx := strings.LastIndex(selector, ":")
					effectiveType = selector[:idx]
					effectiveVersion = selector[idx+1:]
				} else {
					effectiveType = selector
				}
			}

			if effectiveType == "" {
				return fmt.Errorf("kernel type is required")
			}

			// Parse features string into a slice
			var featureList []string
			if features != "" {
				for _, f := range strings.Split(features, ",") {
					f = strings.TrimSpace(f)
					if f != "" {
						featureList = append(featureList, f)
					}
				}
			}

			// jobs flag: only pass if explicitly set
			jobsArg := 0
			if cmd.Flags().Changed("jobs") {
				jobsArg = jobs
			}

			prog := common.NewProgress()
			prog.Start("Pulling kernel...")

			// Build KernelPullInput
			featureStr := strings.Join(featureList, ",")
			kernelInput := inputs.KernelPullInput{
				KernelType:   effectiveType,
				Version:      effectiveVersion,
				Jobs:         jobsArg,
				KeepBuildDir: keepBuildDir,
				CleanBuild:   cleanBuild,
				KernelConfig: kernelConfig,
				SetDefault:   setDefault,
				Features:     featureStr,
			}

			kernelItem, err := kernelAPI.KernelPull(cmd.Context(), kernelInput, func(e event.Progress) {
				if e.Message != "" {
					prog.UpdateText(e.Message)
				}
			})
			prog.Stop()
			if err != nil {
				return err
			}
			if kernelItem != nil {
				common.Cli.Success(
					fmt.Sprintf("Pulled: %s (ID: %s)", kernelItem.Name, common.Cli.FormatID(kernelItem.ID)),
				)
			} else {
				common.Cli.Success("Pull completed")
			}
			return nil
		},
	}

	cmd.Flags().StringVar(&kernelType, "type", "", "Kernel type: firecracker or official")
	cmd.Flags().StringVar(&version, "version", "", "Kernel version")
	cmd.Flags().BoolVarP(&setDefault, "default", "d", false, "Set as default after fetch")
	cmd.Flags().IntVar(&jobs, "jobs", 0, "Parallel build jobs (official only)")
	cmd.Flags().BoolVar(&keepBuildDir, "keep-build-dir", false, "Keep build directory (official only)")
	cmd.Flags().BoolVar(&cleanBuild, "clean-build", false, "Skip cache (official only)")
	cmd.Flags().StringVar(&kernelConfig, "config", "", "Custom kernel config file to apply as a fragment")
	cmd.Flags().StringVar(&features, "features", "", "Comma-separated kernel features (kvm, nftables, tuntap, btrfs)")

	return cmd
}

func newKernelRemoveCmd(kernelAPI api.KernelAPI) *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:               "rm [selectors...]",
		Aliases:           []string{"remove", "delete", "del"},
		Short:             "Remove one or more kernels",
		ValidArgsFunction: completeKernelIDs,
		RunE: func(cmd *cobra.Command, args []string) error {
			if len(args) == 0 {
				return fmt.Errorf("usage error")
			}
			result := kernelAPI.KernelRemove(cmd.Context(), inputs.KernelInput{Identifiers: args, Force: force})
			for _, item := range result.Items {
				if item.IsOK() {
					msg := item.Message
					if msg == "" {
						msg = "Removed"
					}
					common.Cli.Success(msg)
				} else {
					msg := item.Message
					if msg == "" {
						msg = "Remove failed"
					}
					common.Cli.Error(msg)
				}
			}
			if result.HasErrors() {
				return fmt.Errorf("one or more removals failed")
			}
			return nil
		},
	}

	cmd.Flags().BoolVarP(&force, "force", "f", false, "Remove even if referenced by VMs")

	return cmd
}

func newKernelInspectCmd(kernelAPI api.KernelAPI) *cobra.Command {
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:               "inspect [selector]",
		Short:             "Show detailed information about a kernel",
		Args:              cobra.ExactArgs(1),
		ValidArgsFunction: completeKernelIDs,
		RunE: func(cmd *cobra.Command, args []string) error {
			prefix := args[0]

			info, err := kernelAPI.KernelInspect(cmd.Context(), prefix)
			if err != nil {
				return err
			}

			if jsonOutput {
				b, _ := json.MarshalIndent(info, "", "  ")
				fmt.Println(string(b))
				return nil
			}

			name := info.Kernel.Name
			if name == "" {
				name = prefix
			}
			common.Cli.PrintDictTree(common.Cli.ToMap(info), fmt.Sprintf("Kernel: %s", name))
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")

	return cmd
}

func newKernelDefaultCmd(kernelAPI api.KernelAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:               "default [selector]",
		Short:             "Set a kernel as the default",
		Args:              cobra.ExactArgs(1),
		ValidArgsFunction: completeKernelIDs,
		RunE: func(cmd *cobra.Command, args []string) error {
			kernelID, checkErr := common.Cli.CheckArg(cmd, args[0])
			if checkErr != nil {
				return checkErr
			}
			if err := kernelAPI.KernelSetDefault(cmd.Context(), kernelID); err != nil {
				return err
			}
			common.Cli.Success(fmt.Sprintf("Default kernel set to: %s", kernelID))
			return nil
		},
	}

	return cmd
}

func newKernelImportCmd(kernelAPI api.KernelAPI) *cobra.Command {
	var version string
	var setDefault bool

	cmd := &cobra.Command{
		Use:   "import [name] [path]",
		Short: "Register a vmlinux file as a kernel in the database",
		Long: `Register a vmlinux file as a kernel in the database.

Examples:
  mvm kernel import my-kernel ./vmlinux-6.1-x86_64
  mvm kernel import my-kernel ./vmlinux-custom --version 6.1 --default`,
		Args: cobra.ExactArgs(2),
		ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
			if len(args) == 0 {
				return nil, cobra.ShellCompDirectiveNoFileComp // arg0 "name" is new — no completion
			}
			return nil, cobra.ShellCompDirectiveDefault // arg1 "path" — file completion
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			path := args[1]

			// Check source file exists
			if _, err := os.Stat(path); os.IsNotExist(err) {
				return fmt.Errorf("source file not found: %s", path)
			}

			// Build KernelImportInput
			var versionPtr *string
			if version != "" {
				v := version
				versionPtr = &v
			}
			importInput := inputs.KernelImportInput{
				Name:       name,
				Path:       path,
				Version:    versionPtr,
				SetDefault: setDefault,
			}
			kernelItem, err := kernelAPI.KernelImport(cmd.Context(), importInput)
			if err != nil {
				return err
			}
			common.Cli.Success(fmt.Sprintf("Imported: %s", kernelItem.Name))
			common.Cli.Info(fmt.Sprintf("  ID:   %s", common.Cli.FormatID(kernelItem.ID)))
			if setDefault {
				common.Cli.Success(fmt.Sprintf("Default kernel set to: %s", name))
			}
			return nil
		},
	}

	cmd.Flags().StringVar(&version, "version", "", "Override auto-detected kernel version")
	cmd.Flags().BoolVarP(&setDefault, "default", "d", false, "Set as default after import")

	return cmd
}
